### PR TITLE
ZBUG-2192: Fixed IMAP rename folder issue

### PR DESCRIPTION
**Pull request by gabhinavTI**
https://jira.corp.synacor.com/browse/ZBUG-2192

Issue:
Renaming a folder structure in a data source in Zimbra via IMAP results in the below error,
Error:
```
Account "imap1" Failed
system failure: Folder sync failed, system failure: Synchronization of folder 'zimbra_test/a' failed, system failure: Unable to rename the remote folder to zimbra_test/a_2, RENAME failed: RENAME failed 
```

Steps to Reproduce the issue:
- add external account as IMAP(I used Zimbra as an external account ).
- create folder structure in this account: zimbra_test/a/b
- Click on the "update" arrow to sync the account to the server.
- right click on "a" > rename folder: a_1
- click on the "update" arrow to sync the account to the server.
- error message comes up:
- system failure: Folder sync failed, system failure: Synchronization of folder 'zimbra_test/a' failed, system failure: Unable to rename the remote folder to zimbra_test/a_1, RENAME failed: destination folder does exist
- Then there are two folder structures in the account:
/zimbra_test/a
/zimbra_test/a_1/b

Approach and Fix:

- The issue was basically we were renaming the same folder twice and we get the exception and internally folder is created
- The for loop in syncFolder() would pass Folder names in such a way that all parent and child folders gets renamed at a time when they have been renamed
-The code is modified in such a way that only one folder (the last folder) gets renamed at a time if they have been renamed
- In this case, when rename action is called we don't create any extra folder and also the same folder won't get renamed twice
- Now when we try to rename any layer of the folder the rename is successful in UI and DB and also there isn't any extra folder created
-All the emails are syncing up properly as expected in the respective Folders

Testing:
- Unit testing
- Thorough Manual testing from Zimbra web client

### CODE REVIEW

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 1]
```diff
@@ -71,10 +77,42 @@ public void delete() throws IOException {
...
+            connection.rename(path, baseNewPath);
+        }
         return new RemoteFolder(connection, newName);
     }
 
+    // ZBUG-2192
+    // The helper method to make comparison of old and new folder name and
+    // returns boolean value
+    private boolean isSameFolderName(String newPathName) {
+
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:108]: remove empty line here and below, not needed

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 2]
```diff
@@ -233,14 +271,18 @@ public boolean isSelected() {
     public String getPath() {
         return path;
     }
-
+
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:274]: remove extra empty lines here and below, not needed

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
+package com.zimbra.cs.datasource.imap;
```
Comment (dasiyogesh) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:1]: add the license block, refer to other test classes for same.

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 4]
```diff
@@ -71,10 +77,42 @@ public void delete() throws IOException {
...
+                    && StringUtils.substringAfterLast(newName, PATH_SEPARATOR).isEmpty()) {
+                baseNewPath = newName;
+            } else {
+                baseNewPath = StringUtils.substringBeforeLast(baseNewPath, PATH_SEPARATOR) + PATH_SEPARATOR
+                        + StringUtils.substringAfterLast(newName, PATH_SEPARATOR);
+                if (StringUtils.endsWith(baseNewPath, PATH_SEPARATOR) || StringUtils.startsWith(baseNewPath, PATH_SEPARATOR)) {
+                    baseNewPath = StringUtils.removeEnd(baseNewPath, PATH_SEPARATOR);
+                    baseNewPath = StringUtils.removeStart(baseNewPath, PATH_SEPARATOR);
+                }
+            }
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:98]: can you let me know examples of path and newName which meet the condition at line 88, 91 and 94?

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // In this case when middle layer and grand child folder name's are renamed at a
+    // time in this case in
+    // rename method newer path is modified in such a way that only "pq" is renamed
+    // in this loop
+    // an in next zoom Test/Zimbra get's updated to Test/zimbra_1 ref:
```
Comment (k-kato) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:116]: `Test/Zimbra` is `Test/zimbra`, correct?

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // If we update the folder from Test/zimbra/pq to Test/Zimbra_1/pq then the
+    // rename of middle layer folder happens only when loop iterates to Test/Zimbra
```
Comment (k-kato) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:57]: `Test/Zimbra` is `Test/zimbra`, correct?

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    // Below scenario run's in second iteration
+    @Test
+    public void testRenameToMiddleLayerFolder() throws Exception {
```
Comment (k-kato) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:60]: Is the function name correct? I don't think there is any middle layer folder" in originalPath and newPath; parent and child only...?

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 8]
```diff
@@ -71,10 +77,42 @@ public void delete() throws IOException {
 
     public RemoteFolder renameTo(String newName) throws IOException {
         info("renaming folder to '%s'", newName);
-        connection.rename(path, newName);
+
+        // ZBUG-2192
+        // execute only when last folder name's are different
+        // don't change any parent folder names
+        // using "path" as base name we are forming new folder name in such a way only last folder get's renamed at a time
+
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:85]: remove an empty line

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java (hunk 3 above)
Comment (gabhinavTI) [reply store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:17]: Done

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 1 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java]: Done

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java (hunk 5 above)
Comment (gabhinavTI) [reply store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java]: Thanks , it was typo. Fixed it

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java (hunk 7 above)
Comment (gabhinavTI) [reply store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java]: I wanted it to be referenced from another test method , but modified the test name now as  testRenameToDifferentChildFolderName()

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java (hunk 6 above)
Comment (gabhinavTI) [reply store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java]: Fixed

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 9]
```diff
@@ -37,18 +37,24 @@
...
 final class RemoteFolder {
     private final ImapConnection connection;
     private String path;
     private int deleted;
-
+    
+    /**
+     * @see ZMailbox#PATH_SEPARATOR
+     **/
+    public final static String PATH_SEPARATOR = "/";
```
[the remark this answers is not in the document]
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:50]: Had discussed with Stephan on this and he had suggested  unless zmailbox is already a dependency in this class using the Constant from ZMailbox makes no sense and declaring its own constant in current classwould be a good approach

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 4 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:98]: Below are the scenarios for specific condition

- 88 ->When the path is “Test” and newName is “Test_1” (Basically single layer folder name which is modifiable)

- 91 -> When the path is “Test/a/pq” and newName is “Test/a_1/pq_1” (Multi-Layer Folder)

-  94 -> This is basically a check to prevent “/“ getting appended for baseNewPath before or after in any case, check for any unprecedented scenario

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (k-kato) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:50]: I got it.

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 4 above)
Comment (k-kato) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:98]: Thank you for the scenarios. I understood.

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 10]
```diff
@@ -71,10 +77,40 @@ public void delete() throws IOException {
 
     public RemoteFolder renameTo(String newName) throws IOException {
         info("renaming folder to '%s'", newName);
-        connection.rename(path, newName);
+
+        // ZBUG-2192
+        // execute only when last folder name's are different
+        // don't change any parent folder names
+        // using "path" as base name we are forming new folder name 
+        //in such a way only last folder get's renamed at a time
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:85]: nit: add a white space `// in ...`

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 11]
```diff
@@ -240,7 +276,7 @@ public void debug(String fmt, Object... args) {
         }
     }
 
-    public void info(String fmt, Object... args) {
+	public void info(String fmt, Object... args) {
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:279]: It looks just changing white spaces to a tab. White spaces should be used for indentation as same as others. Revert the change so that it does not appear in the code diff.

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
+ * Zimbra Collaboration Suite Server
+ * Copyright (C) 2011, 2013, 2014, 2016 Synacor, Inc.
```
Comment (k-kato) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:4]: `2011, 2013, 2014, 2016` -> 2022

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 13]
```diff
@@ -71,10 +77,40 @@ public void delete() throws IOException {
...
         return new RemoteFolder(connection, newName);
     }
 
+    // ZBUG-2192
+    // The helper method to make comparison of old and new folder name and
+    // returns boolean value
+    private boolean isSameFolderName(String newPathName) {
+        String oldPathName = path.substring(path.lastIndexOf(PATH_SEPARATOR) + 1);
+        newPathName = newPathName.substring(newPathName.lastIndexOf(PATH_SEPARATOR) + 1);
+
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:110]: remove empty line

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
+ * Zimbra Collaboration Suite Server
+ * Copyright (C) 2011, 2013, 2014, 2016, 2022 Synacor, Inc.
```
Comment (dasiyogesh) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:4]: for copyright just keep the year 2022 and remove the others. 
Since this test class is added newly and wasn't present earlier, will just keep the current year and the files which are getting modified there if the current year is not present there we add the current year to existing list.

Comment (gabhinavTI) [reply store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java]: Thanks , wasn't aware of this fixed the Year

**Comment by k-kato** [comment]
Please add `ZBUG-2192: ` to all commit comments.

**Comment by gabhinavTI** [comment]
> Please add `ZBUG-2192: ` to all commit comments.

I'll be squashing the commits, at that time I can give proper commit message and ZBUG number as well

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 15]
```diff
@@ -71,10 +77,39 @@ public void delete() throws IOException {
...
     public RemoteFolder renameTo(String newName) throws IOException {
         info("renaming folder to '%s'", newName);
-        connection.rename(path, newName);
+
+        // ZBUG-2192
+        // execute only when last folder name's are different
+        // don't change any parent folder names
+        // using "path" as base name we are forming new folder name 
+        // in such a way only last folder get's renamed at a time
+        if (!isSameFolderName(newName)) {
```
Comment (log2akshat) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:86]: Avoid negative checks for boolean.

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 16]
```diff
@@ -71,10 +77,39 @@ public void delete() throws IOException {
 
     public RemoteFolder renameTo(String newName) throws IOException {
         info("renaming folder to '%s'", newName);
-        connection.rename(path, newName);
+
+        // ZBUG-2192
+        // execute only when last folder name's are different
+        // don't change any parent folder names
+        // using "path" as base name we are forming new folder name 
+        // in such a way only last folder get's renamed at a time
```
Comment (log2akshat) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:85]: Use multi-line comment style.

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 17]
```diff
@@ -71,10 +77,39 @@ public void delete() throws IOException {
...
+                }
+            }
+            connection.rename(path, baseNewPath);
+        }
         return new RemoteFolder(connection, newName);
     }
 
+    // ZBUG-2192
+    // The helper method to make comparison of old and new folder name and
+    // returns boolean value
```
Comment (log2akshat) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:106]: Use multi-line comment style.

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // In this case when middle layer and grand child folder name's are renamed at a
+    // time in this case in
+    // rename method newer path is modified in such a way that only "pq" is renamed
+    // in this loop
+    // an in next loop Test/zimbra get's updated to Test/zimbra_1     
+    // ref:testRenameToDifferentChildFolderName()
```
Comment (log2akshat) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:133]: Use multi-line comment style.

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // If we update the folder from Test/zimbra/pq to Test/zimbra_1/pq then the
+    // rename of middle layer folder happens only when loop iterates to Test/zimbra
+    // Below scenario run's in second iteration
```
Comment (log2akshat) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:74]: Use multi-line comment style.

File: store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+
+        PowerMock.replay(mockConnection);
+
+        f.renameTo(newPath);
+
+        PowerMock.verify(mockConnection);
+    }
+
+    // In the below test case since grand child name's are same hence rename action
+    // won't be called
```
Comment (log2akshat) [inline store/src/java-test/com/zimbra/cs/datasource/imap/RemoteFolderTest.java:94]: Use multi-line comment style.

**Review by log2akshat — CHANGES_REQUESTED** [review]

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 15 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:86]: @log2akshat Either I add a negative check here or I have to return a negative boolean value from isSameFolderName()
which again is a similar thing. what would be your suggestion to return a negative boolean or continue with the negative check?

Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:86]: @log2akshat, Actually all the if-else conditions are required as there are multiple scenarios in which the Folder name can be passed as an input. Hence in order to handle appropriate way of renaming all the if-else conditions are required.
I tried but couldn't find any other approaches to refactor it, if you have any suggestions we can connect and discuss on that

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 21]
```diff
@@ -71,10 +77,39 @@ public void delete() throws IOException {
...
+        // using "path" as base name we are forming new folder name 
+        // in such a way only last folder get's renamed at a time
+        if (!isSameFolderName(newName)) {
+            String baseNewPath = path;
+            if (StringUtils.substringAfterLast(baseNewPath, PATH_SEPARATOR).isEmpty()
+                    && StringUtils.substringAfterLast(newName, PATH_SEPARATOR).isEmpty()) {
+                baseNewPath = newName;
+            } else {
+                baseNewPath = StringUtils.substringBeforeLast(baseNewPath, PATH_SEPARATOR) + PATH_SEPARATOR
+                        + StringUtils.substringAfterLast(newName, PATH_SEPARATOR);
```
[the remark this answers is not in the document]
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:93]: I have used `concat` operation for this

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 22]
```diff
@@ -71,10 +77,39 @@ public void delete() throws IOException {
...
+                if (StringUtils.endsWith(baseNewPath, PATH_SEPARATOR) || StringUtils.startsWith(baseNewPath, PATH_SEPARATOR)) {
+                    baseNewPath = StringUtils.removeEnd(baseNewPath, PATH_SEPARATOR);
+                    baseNewPath = StringUtils.removeStart(baseNewPath, PATH_SEPARATOR);
```
[the remark this answers is not in the document]
Comment (k-kato) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:96]: another option: use regex to remove (one or more) `/` from the start/end of the string.

[the remark this answers is not in the document]
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:96]: @k-kato I just realized this if block won't be required at all as we can have this check  in all scenarios which ultimately removes the if block at line 94
Using Regex increases the line and complexity of the code as we have to have two different checks at start and end "/" . I guess StringUtils library would be a good option for this. Please let me know about this if shall I proceed as it is or implement the Regex way

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 15 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:86]: @log2akshat for this I have created new method which return the basenewpath and also removed an if condition at L:94 which removed the nested if else block

**Review by log2akshat — APPROVED** [review]

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 15 above)
Comment (dasiyogesh) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:86]: @gabhinavTI suggestion here, make the method `isSameFolderName` other way round as `notSameFolderName` or `folderNameIsChanged` to avoid negative check and from the method return the boolean accordingly.

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 23]
```diff
@@ -71,10 +77,41 @@ public void delete() throws IOException {
...
+     * ZBUG-2192: The helper method to make comparison between old and new folder
+     * name and returns boolean value
+     */
+    private boolean isSameFolderName(String newPathName) {
+        String oldPathName = path.substring(path.lastIndexOf(PATH_SEPARATOR) + 1);
+        newPathName = newPathName.substring(newPathName.lastIndexOf(PATH_SEPARATOR) + 1);
+        return oldPathName.equals(newPathName);
+    }
+    
+    //ZBUG-2192 :Helper method to form and return newpath
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:102]: * add a white space `// ZBUG...`
* fix position of ":"  `// ZBUG-2192: Helper ...`

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 24]
```diff
@@ -71,10 +77,41 @@ public void delete() throws IOException {
...
+            connection.rename(path, createBaseNewPath(newName));
+        }
         return new RemoteFolder(connection, newName);
     }
 
+    /*
+     * ZBUG-2192: The helper method to make comparison between old and new folder
+     * name and returns boolean value
+     */
+    private boolean isSameFolderName(String newPathName) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:96]: add documentation for method here and below

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 25]
```diff
@@ -71,10 +77,41 @@ public void delete() throws IOException {
...
+    //ZBUG-2192 :Helper method to form and return newpath
+    private String createBaseNewPath(String newName) {
+        String baseNewPath = path;
+        if (StringUtils.substringAfterLast(baseNewPath, PATH_SEPARATOR).isEmpty()
+                && StringUtils.substringAfterLast(newName, PATH_SEPARATOR).isEmpty()) {
+            return newName;
+        } else {
+            return StringUtils.substringBeforeLast(baseNewPath, PATH_SEPARATOR).concat(PATH_SEPARATOR)
+                    .concat(StringUtils.substringAfterLast(newName, PATH_SEPARATOR)).replaceAll("^/", "")
+                    .replaceAll("/$", "");
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:111]: If `replaceAll("^/", "")` and `replaceAll("/$", "")` is called for any unprecedented scenario, why don't we add `+` to remove all slashes?
i.e. `replaceAll("^/+", "").replaceAll("/+$", "")`

Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java]: I had missed out scenarios for multiple "/" at before or end . implemented as per your suggestion

**Review by k-kato — APPROVED** [review]

**Review by dasiyogesh — APPROVED** [review]

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java [hunk 26]
```diff
@@ -71,10 +77,50 @@ public void delete() throws IOException {
...
+    private String createBaseNewPath(String newName) {
+        String baseNewPath = path;
+        if (StringUtils.substringAfterLast(baseNewPath, PATH_SEPARATOR).isEmpty()
+                && StringUtils.substringAfterLast(newName, PATH_SEPARATOR).isEmpty()) {
+            return newName;
+        } else {
+            return (StringUtils.substringBeforeLast(baseNewPath, PATH_SEPARATOR).concat(PATH_SEPARATOR)
+                    .concat(StringUtils.substringAfterLast(newName, PATH_SEPARATOR))).replaceAll("^/+", "")
+                    .replaceAll("/+$", "");
+
```
Comment (k-kato) [inline store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java:120]: remove an empty line

**Comment by k-kato** [comment]
> > Please add `ZBUG-2192: ` to all commit comments.
> 
> I'll be squashing the commits, at that time I can give proper commit message and ZBUG number as well

Please do not forget it.

**Comment by gabhinavTI** [comment]
> > > Please add `ZBUG-2192: ` to all commit comments.
> > 
> > 
> > I'll be squashing the commits, at that time I can give proper commit message and ZBUG number as well
> 
> Please do not forget it.

> Added `ZBUG-2192:` for the commit message

File: store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java (hunk 26 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java]: done
